src/math/hit_test.cpp

```diff
@@ -0,0 +1,339 @@
+}
+
+// cubic a(1-t)^3 + 3bt(1-t)^2 + 3c(1-t)t^2 + dt^3
+// becomes
+// At^3 + Bt^2 + Ct + D
+//
+struct CubicCoeff {
+    Point A, B, C, D;
+    
+    // a(1-t)^3 + 3bt(1-t)^2 + 3ct^2(1-t) + dt^3
+    // a - 3at + 3at^2 -  at^3      a(1 - 3t + 3t^2 - t^3)
+    //     3bt - 6bt^2 + 3bt^3     3b(     t - 2t^2 + t^3)
+    //           3ct^2 - 3ct^3     3c(          t^2 - t^3)
+    //                    dt^3      d(                t^3)
+    // ...
+    // D  + Ct  + Bt^2  + At^3
+    //
+    CubicCoeff(Point a, Point b, Point c, Point d) {
+        A = (d - a) + 3.0f*(b - c);
+        B = 3.0f*((c - b) + (a - b));
+        C = 3.0f*(b - a);
+        D = a;
+    }
+    
+    Point eval(float t) const {
+        return ((A*t + B)*t + C)*t + D;
+    }
+};
+
+////////////////////////////////////////////
+
+void HitTester::reset() {
+    m_DW.clear();
+}
+
+void HitTester::reset(const IAABB& clip, FillRule rule) {
+    m_offset = Vec2D{ (float)clip.left, (float)clip.top };
+    m_height = (float)clip.height();
+
+    m_WindingMask = (rule == rive::FillRule::nonZero) ? -1 : 1;
+
+    m_IWidth = clip.width();
+    m_IHeight = clip.height();
+    m_DW.resize(m_IWidth * m_IHeight);
+    for (size_t i = 0; i < m_DW.size(); ++i) {
+        m_DW[i] = 0;
+    }
+
+    m_ExpectsMove = true;
+}
+
+void HitTester::move(Vec2D v) {
+    if (!m_ExpectsMove) {
+        this->close();
+    }
+    m_First = m_Prev = v - m_offset;
+    m_ExpectsMove = false;
+}
+
+void HitTester::line(Vec2D v) {
+    assert(!m_ExpectsMove);
+
+    v = v - m_offset;
+    clip_line(m_height, m_Prev, v, m_DW.data(), m_IWidth);
+    m_Prev = v;
+}
+
+void HitTester::quad(Vec2D b, Vec2D c) {
+    assert(!m_ExpectsMove);
+
+
+    m_Prev = c;
+}
+
+static bool quickRejectCubic(float height, Point a, Point b, Point c, Point d) {
+    const float h = height;
+    return (a.y <= 0 && b.y <= 0 && c.y <= 0 && d.y <= 0) ||
+           (a.y >= h && b.y >= h && c.y >= h && d.y >= h);
+}
+
+struct CubicChop {
+    Vec2D storage[7];
+    
+    CubicChop(Vec2D a, Vec2D b, Vec2D c, Vec2D d) {
+        auto ab = ave(a, b);
+        auto bc = ave(b, c);
+        auto cd = ave(c, d);
+        auto abc = ave(ab, bc);
+        auto bcd = ave(bc, cd);
+        
+        storage[0] = a;
+        storage[1] = ab;
+        storage[2] = abc;
+        storage[3] = ave(abc, bcd);
+        storage[4] = bcd;
+        storage[5] = cd;
+        storage[6] = d;
+    }
+    
+    Vec2D operator[](unsigned index) const {
+        assert(index < 7);
+        return storage[index];
+    }
+};
+
+// Trial and error to pick a good value for this.
+//
+// Subdivision and recursion have their own cost, so at some point
+// just evaluating the cubic (count) times is cheaper than continuing
+// to chop.
+//
+// The key win is quickRejectCubic. This is how we save time over just
+// evaluating the cubic up front.
+//
+#define MAX_LOCAL_SEGMENTS  16
+
+void HitTester::recurse_cubic(Vec2D b, Vec2D c, Vec2D d, int count) {
```

Comment:
This looks a lot like what we do for trim paths in metrics path here: https://github.com/rive-app/rive-cpp/blob/746ac70c45f3983b84611b0aa3d93d0de4885db9/src/shapes/metrics_path.cpp#L83

Would be nice to at some point generalize all of this to be usable in both cases if there is indeed overlap!